### ISSUE TITLE
Merge FFT classes for CPU and CUDA

### ISSUE THREAD

**Issue by ihhub** [ihhub/penguinV#154]
At current stage we have FFT code for CPU which locates in `fft.h` and `fft.cpp` files and CUDA code with `cuda/cuda_fft.cuh` and `cuda/cuda_fft.cu` files respectively.
1. We need to create a base classes for `ComplexData` and `FFTExecutor` to reduce code duplication. Most of the code for CPU and CUDA is common so the reduction of duplication would be obvious.
2. Create new files named `fft_base.h` and `fft_base.cpp` where you put your newly creates 2 base classes. These classes must be within **namespace FFT** as it's done for `fft.h`.

This issue is related to #69

**Comment by MatthewMcGonagle** [comment]
I'm interested in taking a look at this.

**Comment by ihhub** [comment]
Sure.

**Comment by MatthewMcGonagle** [comment]
For functions that need to be declared in the base class but defined in the sub-class, what should the behavior be in the base class? For example, `ComplexData::set()` is called by constructors in both fft classes, but its definition depends on each class. Since it is called by a constructor, we can't make it pure virtual. I was thinking of doing something like `BaseComplexData::set()` is defined to throw `imageException("BaseComplexData::set() should be overridden in sub-class")`. What do you think?

**Comment by MatthewMcGonagle** [comment]
I'm thinking that for `BaseComplexData`, we should actually use a class template based on the type of pointer for `BaseComplexData::data`, and we can use the default constructor to cut out the initializer lists in constructors of the current versions of `ComplexData`. However, the use of stuff like `set()` will still be left in the (to be) subclass constructors.

Do we want pure virtual versions of stuff like `_clean()`, or is it preferrable to have everything decided at compile time?

**Comment by ihhub** [comment]
There is no final decision about pure virtual functions so we could use them. I think you could take a look into `src/memory/memory_allocator.h` as a base class, `src/cuda/cuda_memory.cuh` and `src/opencl/opencl_memory.h` as derived classes. I faced the same situation there. It's not an elegant solution: I moved data type into derived classes because the class has template-based function.

I am okay to make a base FFT class as template. We still need to create a derived a new class to write functions like `_clean`. Just in base class default constructor we shouldn't call any virtual functions like `set` or `_copy`, this logic must be in derived class. (something similar is done in `src/image_buffer.h` and `src/cuda/image_buffer_cuda.cuh`).

**Comment by MatthewMcGonagle** [comment]
Is there a reason that `ComplexData` shouldn't be solved with class template specialization instead of inheritance? Then the template instances of `ComplexDataTemplate` for `kiss_fft_cpx` and `cufftComplex` can be "masked" by using typdefs.

The generic types for `ComplexDataTemplate` could just give a class that throws exceptions.

**Comment by ihhub** [comment]
@MatthewMcGonagle let's do an initial version of merged classes then we could review it together.

**Comment by MatthewMcGonagle** [comment]
I'm writing a simple example to test the changes I have made, but I think I have found a problem with the original ComplexData::get(). In both my code and the master code, if you run something like (`image` has already been properly loaded as a grayscale version of `mercurcy.bmp`):
``` 
 FFT::ComplexData data
 Bitmap_Operation::Save( "grayscale.bmp", image);
 data.set(image);
 image = data.get();
 Bitmap_Operation::Save( "grayscale2.bmp", image);
```
then `grayscale.bmp` is correct, `grayscale2.bmp` is all black.

Now, I think the issue is that in `ComplexData::get()` there is the line
```
out[outY * _width + outX] = static_cast<uint8_t>(_data[inY * _width + inX].r / static_cast<float>(size) + 0.5);
```
inside the for loop. I don't understand why we are dividing by a `static_cast<float>(size)`; here `size` is the total number of pixels in the image. If I replace this line with
```
out[outY * _width + outX] = static_cast<uint8_t>(_data[inY * _width + inX].r + 0.5);
```
then the image is there, but shifted wrapped around the edges. But at least it isn't just black pixels.

**Comment by ihhub** [comment]
I'm just writing unit tests for FFT, First of all image width and height must be same and equal to 2^n. Here is a simple test:
```cpp
// You can load mercury image here and convert to gray scale image
const PenguinV_Image::Image input = Unit_Test::randomImage( 128, 128 ); \\<-- it is very important that width and height are same and equal to 2^n

PenguinV_Image::Image diracDelta( input.width(), input.height() );
diracDelta.fill( 0u );
diracDelta.data()[diracDelta.height() / 2 * diracDelta.rowSize() + diracDelta.width() / 2 ] = 1u;

FFT::ComplexData complexDataInput( input );
FFT::ComplexData complexDataDracDelta( diracDelta );

FFT::FFTExecutor fftExecutor( input.width(), input.height() );

fftExecutor.directTransform( complexDataInput );
fftExecutor.directTransform( complexDataDracDelta );
fftExecutor.complexMultiplication( complexDataInput, complexDataDracDelta, complexDataInput );
fftExecutor.inverseTransform( complexDataInput );

const PenguinV_Image::Image output = complexDataInput.get();
```
In result we should receive totally the same image.

The logic inside `get()` is correct. You have to divide after DFT by number of elements and swap quadrants.

**Comment by ihhub** [comment]
The above example is an implementation of complex multiplication of  function **f()**  with Dirac delta function in frequency domain. The result must be the same after conversion to spatial domain.
kissfft doesn't work properly with any image size. We should fix it later.

**Comment by ihhub** [comment]
@MatthewMcGonagle , here is [an answer](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Properties) (*The inverse transform is given by:*) for your question regarding division by size.

Current code still has some problems like normalization and verification of image size :(

**Comment by MatthewMcGonagle** [comment]
So if I understand your test correctly, you are testing the algebraic properties of DFT and convolution by looking at the specific case of convolution with a dirac delta?

**Comment by ihhub** [comment]
Yes. This test shows that FFT code works perfectly fine on ideal case. You could have a look at filtering.h, GetGaussianKernel function which generates filter. Appling this filter (through FFT) makes your image blur. This is more use case for FFT but it's hard to write a correct unit test.

**Comment by MatthewMcGonagle** [comment]
Trying to setup an example to check the results for Cuda based fft, and I'm looking at `examples/cuda/example_image_function.cpp`. For the function `gpuBased()`, I don't understand why the following assignment is to `Image` instead of `ImageCuda`:

```
 PenguinV_Image::Image image = Image_Function_Cuda::ConvertToCuda( Bitmap_Operation::Load( "mercury.bmp" ) );
```
We are passing the instance of `ImageCuda` returned by `ConvertToCuda()` to `Image::operator=()`, which is okay as `ImageCuda` is derived from `Image`. However, the end result is of type `Image` instead of `ImageCuda`? Shouldn't the example be using `ImageCuda`?

**Comment by MatthewMcGonagle** [comment]
Another question: When I use `FFT_Cuda::ComplexData::set(Image &)` on something of type `Image` instead of  type `ImageCuda`, I get the exception "Cannot create FFT plan on CUDA device". So it seems doing so gives me an error related to calling `cufftPlan2d()`. Is this expected to happen?

**Comment by ihhub** [comment]
> Trying to setup an example to check the results for Cuda based fft, and I'm looking at `examples/cuda/example_image_function.cpp`. For the function `gpuBased()`, I don't understand why the following assignment is to `Image` instead of `ImageCuda`:
> 
> ```
>  PenguinV_Image::Image image = Image_Function_Cuda::ConvertToCuda( Bitmap_Operation::Load( "mercury.bmp" ) );
> ```
> We are passing the instance of `ImageCuda` returned by `ConvertToCuda()` to `Image::operator=()`, which is okay as `ImageCuda` is derived from `Image`. However, the end result is of type `Image` instead of `ImageCuda`? Shouldn't the example be using `ImageCuda`?

Nope. All image functions accept base type `Image`. If you follow the code each function calls `Image_Function_Helper` namespace function which creates (if necessary) a proper instance of Image class. Going forward with recent changes we could have a single interface for any image types (as it's done in `penguinv.cpp` file) so we could just create an image type which we want to work with and we don't need to change function names after :)

**Comment by ihhub** [comment]
If you are not sure about the correct types of Image passed into your function you could check like this:
```cpp
if (image.type != Image_Cuda().type())
```

**Comment by ihhub** [comment]
> Another question: When I use `FFT_Cuda::ComplexData::set(Image &)` on something of type `Image` instead of type `ImageCuda`, I get the exception "Cannot create FFT plan on CUDA device". The error is caused by calling `FFT_Cuda::FFTExecutor::intialize()`. So it seems doing so gives me an error related to calling `cufftPlan2d()`. Is this expected to happen?

Could you please paste you code? I'm not sure about this.

**Comment by MatthewMcGonagle** [comment]
Here's a minimal example with error (and passes with using `ImageCuda`); I think the `FFTExecutor::initialize()` might be a bit of a red herring. If I run it without using `FFTExecutor::intialize()`, inside `makeError()` everything runs through until the "End of main()" output, but I then get a Microsoft Visual C++ Runtime Library Error (I'm building on windows). However, if I just run `doErrorFree()` then everything is okay.

```
// Example of application of library utilization for fft.

#include <iostream>
#include "../../../src/image_buffer.h"
#include "../../../src/image_function.h"
#include "../../../src/FileOperation/bitmap.h"

#include "../../../src/cuda/cuda_fft.cuh"
#include "../../../src/cuda/cuda_device.cuh"
#include "../../../src/cuda/cuda_helper.cuh"
#include "../../../src/cuda/image_buffer_cuda.cuh"
#include "../../../src/cuda/image_function_cuda.cuh"

void doErrorFree();
void makeError();

int main() {

    multiCuda::CudaDeviceManager::instance().initializeDevices();

    // It is recommended to use preallocated buffers for GPU memory usage
    // So we preallocate 32 MB of GPU memory for our usage

    multiCuda::MemoryManager::memory().reserve( 32 * 1024 * 1024 );

    try {
        doErrorFree();
    }
    catch (const std::exception & except) {
        std::cout << "Exception: " << except.what() << std::endl;
    }
    std::cout << "Finished Error Free" << std::endl;

    try {
        makeError();
    }
    catch (const std::exception & except) {
        std::cout << "Exception: " << except.what() << std::endl;
    }

    std::cout << "End of main()" << std::endl;
    return 0;
}

void doErrorFree() {

    PenguinV_Image::Image image;

    // The file mercury.bmp needs to be located in the directory the executable is built to.
    image = Bitmap_Operation::Load( "mercury.bmp" );

    // If the image is empty it means that the image doesn't exist or the file is not readable
    if( image.empty() )
        throw imageException( "Cannot load the image" );

    // Convert to gray-scale image if it's not
    if( image.colorCount() != PenguinV_Image::GRAY_SCALE )
        image = Image_Function::ConvertToGrayScale( image );

    FFT_Cuda::ComplexData data(Image_Function_Cuda::ConvertToCuda(image) );

    FFT_Cuda::FFTExecutor fft;
    fft.initialize(data.width(), data.height());

}

void makeError() {

    PenguinV_Image::Image image;

    // The file mercury.bmp needs to be located in the directory the executable is built to.
    image = Bitmap_Operation::Load( "mercury.bmp" );

    // If the image is empty it means that the image doesn't exist or the file is not readable
    if( image.empty() )
        throw imageException( "Cannot load the image" );

    // Convert to gray-scale image if it's not
    if( image.colorCount() != PenguinV_Image::GRAY_SCALE )
        image = Image_Function::ConvertToGrayScale( image );

    FFT_Cuda::ComplexData data(image);

    FFT_Cuda::FFTExecutor fft;
    fft.initialize(data.width(), data.height());
}
```

**Comment by MatthewMcGonagle** [comment]
I think the issue might be with `FFT_CUDA::ComplexData::set(Image &)` calling `copyFromImageCuda()`. If I understand this correctly, this is working with addresses on the GPU and not on the host machine. So passing in the pointer from a regular `Image` is a pointer to local memory instead of memory on the GPU?

**Comment by ihhub** [comment]
set function of ComplexData indeed accepts base type Image because as I mentioned earlier all image functions do the same. In error free case you convert CPU memory based image into GPU memory based image before passing into set function. In error case you don't. In image_function example for CUDA we do conversion as well.

About Image class itself: base class surprisingly knows everything about every child class via registration of static functions inside singleton class which locates in the body of Image class. All base operaions as copy, allocate, set and dealllocate go through this singleton class which calls registered static functions based on the type of Image. In this case even if you make a base class but change the internal type the Image would behave differently.

Such technique is done to support any possible platform as CPU, OpenCL and CUDa. Going forward (as I wrote previously) we have penguinv file which contains a set of funxtions similar like any other. The magic of these functions is that based on input type of the Image the image processing is different. For example, you pass CPU memory based image so your processing is done on CPU, you pass CUDA based image - processing on GPU. The only difference is how you allocate your input images.

**Comment by MatthewMcGonagle** [comment]
I see now that `Image` and `ImageCuda` are accessing functions that are statically stored in `Function::Facade`, but the class `FFT_Cuda::ComplexData` still seem to be making memory allocations/deallocations independent of what the `Image` instance is doing.

To start with, my understanding is that the four `__global__` functions the beginning of `fft_cuda.cu` are run on the GPU, and any memory locations they refer to are on the GPU. Also none of them use any of the internals of `Image` or `ImageCuda`. Is my understanding correct?

However, `FFT_Cuda::ComplexData::set(Image &)` is launching `__global__ copyFromImageCuda()` regardless of how the image actually behaves internally. Also, in the definition of `FFT_Cuda::ComplexData::get()`, the function constructs a regular `Image` object (NOT `ImageCuda`) and then uses it with `__global__ copyToImageCuda()`, which will also assume everything is happening on the GPU. The problem here is that the regular `Image` is on the host machine, not the GPU?

**Comment by ihhub** [comment]
You're correct regarding 4 first functions at beginning of the file. These functions only work with GPU memory and executed only on GPU.

Adding a type check inside `set` function would solve the problem with Image type. `set` function indeed has an error, it should be `ImageCuda`.

If this wouldn't solve the problem, give me some time to check the behaviour myself.

**Comment by ihhub** [comment]
I [fixed](https://github.com/ihhub/penguinV/commit/8dc82a9e8649d1b107a27294cc2a161418ebe82a) the source of black images. The problem was in wrong input parameters for `memcpy` function on CUDA (input and output pointers were swapped). Now you could try your example code with this fix. It should work now.

**Comment by ihhub** [comment]
Last thing: please use image with dimensions equal to 2^n such as 256 x 256 or 128 to 128. There is a modification of your code to test properly:
```cpp
void cudaFFTExample()
{
    // First thing we should check whether the system contains CUDA device
    if( !multiCuda::isCudaSupported() ) 
        throw imageException("CUDA device is not found in current system.");

    multiCuda::CudaDeviceManager::instance().initializeDevices();

    // It is recommended to use preallocated buffers for GPU memory usage
    // So we preallocate 32 MB of GPU memory for our usage

    multiCuda::MemoryManager::memory().reserve( 32 * 1024 * 1024 );

    // The file mercury.bmp needs to be located in the directory the executable is built to.
    PenguinV_Image::ImageCuda image = Image_Function_Cuda::ConvertToCuda( Bitmap_Operation::Load( "mercury.bmp" ));

    // If the image is empty it means that the image doesn't exist or the file is not readable
    if( image.empty() )
        throw imageException( "Cannot load the image" );

    // Convert to gray-scale image if it's not
    if( image.colorCount() != PenguinV_Image::GRAY_SCALE )
        image = Image_Function_Cuda::ConvertToCuda(Image_Function_Cuda::ConvertToGrayScale( image ));

    // Save a copy of .bmp to check that the ImageCuda version looks good.
    Bitmap_Operation::Save( "cuda_input.bmp", Image_Function_Cuda::ConvertFromCuda( image )); 

    // Set up the data and save a picture of the data to see that everything worked okay.
    FFT_Cuda::ComplexData data( image );
    //saveDataPic("cudafft1.bmp", data);

    PenguinV_Image::Image diracDeltaCPU( image.width(), image.height() );
    diracDeltaCPU.fill( 0u );
    diracDeltaCPU.data()[diracDeltaCPU.height() / 2 * diracDeltaCPU.rowSize() + diracDeltaCPU.width() / 2 ] = 1u;
    PenguinV_Image::Image diracDelta = Image_Function_Cuda::ConvertToCuda( diracDeltaCPU );

    FFT_Cuda::ComplexData complexDataDracDelta( diracDelta );

    // Do the direct fft and save a picture of the result. 
    FFT_Cuda::FFTExecutor fft( data.width(), data.height() );

    fft.directTransform( data );
    fft.directTransform( complexDataDracDelta );
    fft.complexMultiplication( data, complexDataDracDelta, data );
    fft.inverseTransform( data );

    Bitmap_Operation::Save("cuda_output.bmp", Image_Function_Cuda::ConvertFromCuda( data.get() ));
}
```

I resize in image editor `mercury,bmp` to 256 by 256 pixels before running your test code. If you don't do complex multiplication (your original code) when quadrants in output image would be swapped.

**Comment by MatthewMcGonagle** [comment]
The [cudaFFT documentation](https://docs.nvidia.com/cuda/cufft/index.html#multiple-GPU-cufft-supported-functionality) says that for 2D C2C the dimensions should factor into prime less than 127 (the smaller the primes the better). Is the 2^n format really necessary for `FFT_Cuda::FFTExecutor` and `FFT_Cuda::ComplexData`?

**Comment by ihhub** [comment]
Not really, actually. 2^n gives the best performance.

**Comment by ihhub** [comment]
Hi, @MatthewMcGonagle . How is it going? :)

**Comment by MatthewMcGonagle** [comment]
Hi @ihhub, I've tried out your example code, and I enlarged `mercury.bmp` to be 256x256 and everything with the original code is working well. However, with the original `mercury.bmp` dimensions, the output isn't quite right. However, it does all work well in the 256x256 case.

I'll get to work on finishing the base classes. Thanks for the help.

**Comment by MatthewMcGonagle** [comment]
A brief question as to how to resolve an inconsistency in the two versions of `FFTExecutor`. The non-cuda version `FFT::FFTExecutor::directTransform( const ComplexData & in, ComplexData & out )` has its first parameter as constant and the cuda version `FFT_Cuda::FFTExecutor::directTransform(ComplexData & in, ComplexData & out )` does not. If you try to make the cuda version have a constant first parameter then there appears to be incompatibility issues with the function calls to the cuda fft library. 

What seems like the best way to resolve this?

**Comment by ihhub** [comment]
Logically, `cufftExecC2C` must not change input array but there is no guarantee as we don't know the code inside the function. Current solution would be to make both `FFT` and `FFT_Cuda` to use non-const reference. For CUDA we could change it later to const reference with `const_cast` inside the function (which is unsafe of course).

**Comment by MatthewMcGonagle** [comment]
On the other hand, there doesn't really seem to be any space saved by putting `FFTExecutor::directTransform` and `FFTExecutor::inverseTransform` in the base class. You still have to implement the two parameter versions, e.g. `FFTExecutor::directTransform(ComplexData & in, ComplexData & out)` in the sub-class. The overloaded single parameter versions still need to be put into the sub-class with a `using` statement, but their definition is so small this isn't really saving anything.

In fact it makes everything more complicated as including them in the base class forces you to make a base class template. Without the base class knowing about `directTransform` and `indirectTransform`, the base class can just be an abstract class, no need for templates.

Should I just leave them `directTransform` and `inverseTransform` out of the base class?

**Comment by ihhub** [comment]
No, you could put these functions in derived classes, no harm in this.

**Comment by MatthewMcGonagle** [comment]
@ihhub Do you have a preference as to how I'm splitting the base implemenations across files? If not, I'll also make alterations to the make files/visual studio solutions for running unit tests.

**Comment by ihhub** [comment]
> I think my changes are finished; I just need to verify everything works with the unit tests. However, I have one last question.
> 
> The file `base_fft.h` uses a class template and so the definitions of this class template need to be included with `base_fft.h`. However, to keep the file from being too long, I put the definitions in a file `base_fft.tpp`, note the "t" in `.tpp`, and this file is included at the end of `base_fft.h`.
> 
> Is this acceptable? Or would you rather the definitions be put in `base_fft.h`? If it is okay, is the extension `.tpp` okay or do you have another extension you would prefer?
> 
> Edit: Another issue that comes up when looking at doing unit tests: both of the classes in `base_fft.h` inherit from a `Rectangle` class to deal with checks of width and height. Right now I have the definitions of `Rectangle` in `base_fft.cpp` which means the makefiles/visual studio files for the unit tests would need to be updated. Or should I try to deal with this another way?

@MatthewMcGonagle sorry, I missed your previous message. If it's a template class then just put definition of the class within `*.h` file. Why do you need `Rectangle` class? Pleas have a look into `src/parameter_validation.h` file. This file contains functions just for size validation.

**Comment by ihhub** [comment]
> @ihhub Do you have a preference as to how I'm splitting the base implemenations across files? If not, I'll also make alterations to the make files/visual studio solutions for running unit tests. Then everything should be ready for a pull request.

We could put `base_fft.h` file inside `src/math` directory.

**Comment by ihhub** [comment]
From my own experience Rectangle class is not suitable name and no need to do this. We could make public function inside ComplexData and call it from everywhere we want. I'm not really sure how checks look like. Let's review them in pull request. It's easier to see :)

**Comment by ihhub** [comment]
Yes, you have to put .cpp file into solution files for proper compilation.

**Comment by ihhub** [comment]
@MatthewMcGonagle how is it going? I though it's already done :)

**Comment by ihhub** [comment]
@MatthewMcGonagle don't worry about an amount of commits. We squash all commits into one during merging.

**Comment by MatthewMcGonagle** [comment]
@ihhub I've already recloned and added the changed files in to shorten the commits. I'm just rerunning the tests before I submit a pull request.

One issue that I've come up with while getting the cuda examples to compile is that the README.md inside src/cuda could use some instructions on properly setting up a visual studio solution to compile and link the source files inside src/cuda. For example, make sure to add `cudart.lib` and `cufft.lib` to the linker's additional dependencies, and if one is using CUDA 10.0, then make sure the project is compiling as x64.

**Comment by ihhub** [comment]
That's okay. I have CUDA 8 and I cannot compile for 32-bit as well.
